### ISSUE TITLE
Types not being resolved when using baseURL

### ISSUE THREAD

**Issue by briandipalma** [microsoft/TypeScript#31536]
## Search Terms

baseUrl, typeRoots, traceResolution

## Suggestion

`typeRoots` should take into account `baseUrl`?

## Use Cases

The issue can be seen in this repository https://github.com/briandipalma/test-symlinked-ts-packages

We rely on linked packages during development so we'd often have packages with no `node_modules` symlinked into an application. We can fix most errors that causes by setting `baseUrl` to the application's `node_module` which would contain all the packages the compiler needs but one set of errors we haven't been able to fix. Those are declaration file errors for thirdparty packages such as React etc. We have them in the app `node_modules` directory but the TS compiler doesn't seem to pick them up even when we use `typeRoots`.

These are the sort of errors we get:

```
packages/test-package/test.ts:1:19 - error TS7016: Could not find a declaration file for module 'react'. '/Users/briand/dev/test-symlinked-ts-packages/apps/test-app/node_modules/react/index.js' implicitly has an 'any' type.
  Try `npm install @types/react` if it exists or add a new declaration (.d.ts) file containing `declare module 'react';`

1 import React from "react";
                    ~~~~~~~


Found 1 error.
```

And yet `@types/react` is present in the application `node_modules` folder.
Obviously if the packages are installed in the symlinked package it all works but this would result in our application having multiple copies of packages like React

Honestly it almost seems like a bug to me. I also don't understand why `traceResolution` doesn't cover looking for declaration files...

**Comment by RyanCavanaugh** [comment]
> typeRoots should take into account baseUrl

These are two unrelated things, really. Type roots do not affect module resolution at all; this is intentional.

**Comment by briandipalma** [comment]
The issues isn't with resolving modules, those are resolved fine. It's around not resolving the declaration files even when `typeRoots` is configured to point to the correct `node_modules/@types` folder

**Comment by briandipalma** [comment]
Inside

`function resolveExternalModule(location, moduleReference, moduleNotFoundError, errorNode, isForAugmentation) {`

The line

`var resolvedModule = ts.getResolvedModule(currentSourceFile, moduleReference);`

returns two different values for `moduleReference` `"react"`. When the package has a local `node_modules` the package ID is `@types/react` when there is no `node_modules` the package ID is simply the `react` package...

So basically `sourceFile.resolvedModules` is being populated with different values depending on if `baseUrl` is set.

**Comment by briandipalma** [comment]
Inside

`function processImportedModules(file) {`

The line

`var resolutions = resolveModuleNamesReusingOldState(moduleNames, ts.getNormalizedAbsolutePath(file.originalFileName, currentDirectory), file);`

is where the two different resolutions are created.

I'll pick this up again when I have some free time, I'll have to park it for the moment.

`
 if (file.fileName.includes("test-symlinked-ts-packages/packages/test-package/test.ts")) {
                    debugger;
                }
`

in the above function helps set up the debugging session.

**Comment by juanger** [comment]
@briandipalma From what I've read in other issues, this is the intended behavior. You have to set your `paths` config like this:

```
"paths": {
    "*": ["@types/*", "*"]
}
```

 See: https://github.com/microsoft/TypeScript/issues/14551#issue-212916481
